### PR TITLE
add evm_unlockUnknownAccount

### PR DESCRIPTION
**Pull request by rymnc**
This feature is much-needed for on the fly transaction creation, I hope you accept this PR!
I can write more tests if required

### CODE REVIEW

**Comment by davidmurdoch** [comment]
Thanks for the PR. We'll need to discuss implementation details (such as the RPC endpoint) before mergeing.

Related issue: #588

**Comment by aaryamannchallani** [comment]
> Thanks for the PR. We'll need to discuss implementation details (such as the RPC endpoint) before mergeing.
> 
> Related issue: #588

Thank You!

**Comment by aaryamannchallani** [comment]
@davidmurdoch would you be open to using the
`web3.extend` module?
It could help if you are planning of extending the json rpc api
[Documentation](https://web3js.readthedocs.io/en/v1.2.0/web3-eth.html#extend)
It has i/o formatters as well

**Comment by davidmurdoch** [comment]
I don't think this is something that could be done from within ganache-core, as we don't use web3 (except for a couple of RPC calls when in forking mode). Maybe trufflesuite/truffle could extend their injected web3 with ganache's missing RPC methods, if they don't already.

**Comment by aaryamannchallani** [comment]
yes, that's true. Just had a little doubt though. does truffle use ganache-core?

**Comment by davidmurdoch** [comment]
Yes, it does for the default development chain.

**Comment by aaryamannchallani** [comment]
@davidmurdoch any update on integrating this endpoint into the suite?

**Comment by davidmurdoch** [comment]
@aaryamannchallani We're discussing this internally. For now, let's change the namespace to `evm_unlockAccount`, as we use that for our non-standard rpc endpoints.

There are some security considerations to be aware of here. As is, this will allow any user to unlock any account, even an intentionally locked account (accounts that are in `accounts` but not in `unlocked_accounts`, which can happen at start up or via some of the `personal_` RPC calls).

So we need to a think of a way to secure this RPC endpoint by default, as ganache is often used in collaborative environments.

**Comment by aaryamannchallani** [comment]
got it, will make that change
as for the locked accounts, I'll have to read the codebase a bit more

File: lib/subproviders/geth_api_double.js
Comment (davidmurdoch) [inline lib/subproviders/geth_api_double.js:586]: Let's make this should return `false` instead of `"Account Already Unlocked"`.

File: lib/subproviders/geth_api_double.js
Comment (davidmurdoch) [inline lib/subproviders/geth_api_double.js:578]: We chatted about this internally and decided on this method should be named `evm_unlockUnknownAccount`.

File: lib/subproviders/geth_api_double.js
Comment (davidmurdoch) [inline lib/subproviders/geth_api_double.js:582]: Before we can merged this the behavior needs to be changed so that it _does not_ unlock accounts found in `this.state.personal_accounts`, as these can already be unlocked (if previously locked at start up or via `personal_lockAccount`) via `personal_unlockAccount`.

Comment (davidmurdoch) [reply lib/subproviders/geth_api_double.js:582]: In the case where the account cannot be unlocked, return an error in the callback:

```
  callback(new Error("cannot unlock known/personal account"));
```

**Review by davidmurdoch — CHANGES_REQUESTED** [review]
We chatted about this internally and decided on this method should be named `evm_unlockUnknownAccount`.

Before we can merge this the behavior needs to be changed so that it _does not_ unlock accounts found in `this.state.personal_accounts`, as these can already be unlocked (if previously locked at start up or via `personal_lockAccount`) via `personal_unlockAccount`.

File: lib/subproviders/geth_api_double.js
Comment (aaryamannchallani) [reply lib/subproviders/geth_api_double.js]: Resolved in this [commit](https://github.com/trufflesuite/ganache-core/pull/622/commits/010e21e05210423bcce925c8fdd7167745a8e2d0)

**Comment by aaryamannchallani** [comment]
@davidmurdoch let me know if this is what you require!

**Comment by davidmurdoch** [comment]
Can you enable "allow edits from maintainers" on this PR so I can commits some additional work directly to this PR?

**Comment by aaryamannchallani** [comment]
It is enabled!

**Comment by aaryamannchallani** [comment]
ah thank you!
will it be in the next version 🙂

**Comment by davidmurdoch** [comment]
I didn't mean to close this just yet! I just pushed the wrong local branch :-). I'm working on fixing it back up now.

**Comment by aaryamannchallani** [comment]
ah okay cool

**Comment by davidmurdoch** [comment]
@aaryamannchallani Merged!

**Comment by aaryamannchallani** [comment]
Awesome! let me know if I can help anywhere else

**Comment by surajsjain** [comment]
How do I use this now for unlocking all the accounts from a forked chain?

**Comment by davidmurdoch** [comment]
You'll still need to unlock them individually before using each one. We don't yet have a way of unlocking _all_ accounts.

```javascript
// unlock an address
await provider.send("evm_addAccount", [address, passphrase] );
await provider.send("personal_unlockAccount", [address, passphrase] );
```

**Comment by MicaiahReid** [comment]
To add some more detail for any future readers, see #3603

**Pull request closed without merging**
